### PR TITLE
pages: implement basic ElasticSearch UI

### PR DESCRIPTION
**Pull request by Xe**
Replaces #271.

This implements a basic/hacky search UI that will take things from either the URL bar or input box, query ElasticSearch (debouncing up to 150 ms of inactivity), and then present those results on a page. This is not elegant, but it does work.

### CODE REVIEW

File: pages/api/search.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+export interface SearchResponse {
+  status: string;
+  error?: string;
+  query: string;
+  data: SearchResult[];
+}
+
+export default async function handler(req, resp) {
```
Comment (tylersmalley) [inline pages/api/search.ts:51]: I would recommend the search from https://github.com/tailscale-dev/tailscale-dev/pull/269#issue-1810851839

**Comment by tylersmalley** [comment]
Tried searching but getting an error

image

File: pages/search.tsx [hunk 2]
```diff
@@ -0,0 +1,154 @@
+import Image from '@/components/image';
+import Link from 'next/link';
+import { PageSEO } from '@/components/seo';
+import { useEffect, useState } from 'react';
+import { SearchResponse, SearchResult } from './api/search';
+import NoSSRWrapper from '@/components/no-ssr-wrapper';
+
+const SearchResult = (item: SearchResult) => {
```
Comment (tylersmalley) [inline pages/search.tsx:8]: This is overloading the import from `./api/search`.

File: pages/search.tsx [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            <>
+              <Spinner />
+            </>
+          )}
+          <p>
+            {searchTerm === '' && <span>Enter some terms to search!</span>}
+            {searchResults.length === 0 && searchTerm !== '' && !loading && (
+              <span>No results found for {searchTerm}</span>
+            )}
+            <ul>{searchResults.map((item) => SearchResult(item))}</ul>
```
Comment (tylersmalley) [inline pages/search.tsx:148]:
```suggestion
            <ul>{searchResults.map((item) => <SearchResult {...item} />)}</ul>
```

**Comment by tylersmalley** [comment]
Couple UX suggestions:

1. Persist the search term in the search input in the header, and remove the additional input on the search results page.

It's an odd behavior to search for something by typing into the input in the header to have it move.

image

2. Remove the search box on `/blog`. It's pretty confusing to see two separate search boxes with different behavior.

File: pages/search.tsx (hunk 3 above)
Comment (tylersmalley) [inline pages/search.tsx:148]: Can we use the `ListItem` component here?

File: pages/search.tsx [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+                  <>
+                    {SearchHit(item)}
+                    {/* {ListItem({
+                    slug: item.id,
+                    path: item.permalink,
+                    title: item.title,
+                    date: item.date,
+                    summary: item.description || item.summary,
+                    tags: item.tags,
+                  })} */}
```
Comment (Xe) [inline pages/search.tsx:172]:
```suggestion
```

**Review by tylersmalley — APPROVED** [review]
LGTM

**Pull request merged**